File: java/jelly/JellyShape.java
package jelly;

import java.util.*;

public class JellyShape {
    private static final Map<List<Point>, JellyShape> shapes = new HashMap<>();

    public static synchronized JellyShape registerShape(List<Point> positions) {
        List<Point> sorted = new ArrayList<>(positions);
        Collections.sort(sorted);
        // 不変リストにしてキーとする
        List<Point> key = List.copyOf(sorted);
        if (!shapes.containsKey(key)) {
            shapes.put(key, new JellyShape(key));
        }
        return shapes.get(key);
    }

    private final List<Point> positions;
    private final int w;
    private final int h;
    private final int[] lines;
    private final int[] adjacentLines;

    private JellyShape(List<Point> positions) {
        this.positions = positions;
        int maxW = 0;
        int maxH = 0;
        for (Point p : positions) {
            if (p.x() > maxW) maxW = p.x();
            if (p.y() > maxH) maxH = p.y();
        }
        this.w = maxW + 1;
        this.h = maxH + 1;

        int[] lines = new int[this.h];
        for (Point p : positions) {
            lines[p.y()] |= 1 << p.x();
        }
        this.lines = lines;

        int[] adj = new int[this.h + 2];
        for (int i = 0; i < lines.length; i++) {
            int line = lines[i];
            adj[i] |= line << 1;
            adj[i + 1] |= line | (line << 2);
            adj[i + 2] |= line << 1;
        }
        this.adjacentLines = adj;
    }

    public List<Point> getPositions() {
        return positions;
    }

    public int getW() {
        return w;
    }

    public int getH() {
        return h;
    }

    public int[] getLines() {
        return lines;
    }

    public int[] getAdjacentLines() {
        return adjacentLines;
    }

    public boolean occupyPosition(int x, int y) {
        for (Point p : positions) {
            if (p.x() == x && p.y() == y) {
                return true;
            }
        }
        return false;
    }

    public JellyShape concatenated(JellyShape other, int dx, int dy) {
        List<Point> newPositions = new ArrayList<>();
        int offsetX = 0;
        int offsetY = 0;
        if (dx < 0) {
            offsetX = -dx;
            dx = 0;
        }
        if (dy < 0) {
            offsetY = -dy;
            dy = 0;
        }

        for (Point p : this.positions) {
            newPositions.add(new Point(p.x() + offsetX, p.y() + offsetY));
        }
        for (Point p : other.positions) {
            newPositions.add(new Point(p.x() + dx, p.y() + dy));
        }
        return registerShape(newPositions);
    }

    @Override
    public String toString() {
        return "JellyShape{w=" + w + ", h=" + h + ", positions=" + positions + "}";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        JellyShape that = (JellyShape) o;
        return Objects.equals(positions, that.positions);
    }

    @Override
    public int hashCode() {
        return Objects.hash(positions);
    }
}

File: rust/src/jelly_shape.rs
use std::sync::{Arc, OnceLock};
use crate::point::Point;

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct JellyShape {
    inner: Arc<JellyShapeInner>,
}

#[derive(Debug, PartialEq, Eq, Hash)]
struct JellyShapeInner {
    positions: Vec<Point>,
    w: i32,
    h: i32,
    lines: Vec<i32>,
    adjacent_lines: Vec<i32>,
}

static SHAPES: OnceLock<dashmap::DashMap<Vec<Point>, JellyShape>> = OnceLock::new();

fn get_shapes() -> &'static dashmap::DashMap<Vec<Point>, JellyShape> {
    SHAPES.get_or_init(|| dashmap::DashMap::new())
}

impl JellyShape {
    pub fn register_shape(positions: Vec<Point>) -> Self {
        let mut sorted = positions;
        sorted.sort();

        let shapes = get_shapes();

        if let Some(shape) = shapes.get(&sorted) {
            return shape.value().clone();
        }

        let mut max_w = 0;
        let mut max_h = 0;
        for p in &sorted {
            if p.x > max_w { max_w = p.x; }
            if p.y > max_h { max_h = p.y; }
        }
        let w = max_w + 1;
        let h = max_h + 1;

        let mut lines = vec![0; h as usize];
        for p in &sorted {
            lines[p.y as usize] |= 1 << p.x;
        }

        let mut adj = vec![0; (h + 2) as usize];
        for i in 0..lines.len() {
            let line = lines[i];
            adj[i] |= line << 1;
            adj[i + 1] |= line | (line << 2);
            adj[i + 2] |= line << 1;
        }

        let shape = JellyShape {
            inner: Arc::new(JellyShapeInner {
                positions: sorted.clone(),
                w,
                h,
                lines,
                adjacent_lines: adj,
            }),
        };

        shapes.insert(sorted, shape.clone());
        shape
    }

    pub fn positions(&self) -> &[Point] {
        &self.inner.positions
    }

    pub fn w(&self) -> i32 {
        self.inner.w
    }

    pub fn h(&self) -> i32 {
        self.inner.h
    }

    pub fn lines(&self) -> &[i32] {
        &self.inner.lines
    }

    pub fn adjacent_lines(&self) -> &[i32] {
        &self.inner.adjacent_lines
    }

    pub fn occupy_position(&self, x: i32, y: i32) -> bool {
        self.inner.positions.iter().any(|p| p.x == x && p.y == y)
    }

    pub fn concatenated(&self, other: &Self, mut dx: i32, mut dy: i32) -> Self {
        let mut new_positions = Vec::new();
        let mut offset_x = 0;
        let mut offset_y = 0;
        if dx < 0 {
            offset_x = -dx;
            dx = 0;
        }
        if dy < 0 {
            offset_y = -dy;
            dy = 0;
        }

        for p in &self.inner.positions {
            new_positions.push(Point::new(p.x + offset_x, p.y + offset_y));
        }
        for p in &other.inner.positions {
            new_positions.push(Point::new(p.x + dx, p.y + dy));
        }
        Self::register_shape(new_positions)
    }
}
